File: LeetCode/Algorithms/0003 Longest Substring Without Repeating Characters.cc
class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    if (s.empty()) return 0;
    array<int, 128> arr;
    arr.fill(-1);
    int max_len = 0;
    int current_len, char_code;
    for (int i = 0, j = 0; j < s.size(); j++) {
      char_code = static_cast<int>(s[j]);
      if (arr[char_code] != -1) {
        i = max(i, 1 + arr[char_code]);
      }
      current_len = j - i + 1;
      max_len = max(max_len, current_len);
      arr[char_code] = j;
    }
    return max_len;
  }
};

File: LeetCode/Algorithm/go/0003-Longest Substring Without Repeating Characters.go
package main

func lengthOfLongestSubstring(s string) int {
	start, max := 0, 0
	var checked [256]int
	for i, c := range s {
		if checked[c] > start {
			start = checked[c]
		}

		if i-start+1 > max {
			max = i - start + 1
		}

		checked[c] = i + 1
	}

	return max
}

func main() {
	lengthOfLongestSubstring(" ")
}
